language/move-prover/bytecode-to-boogie/src/env.rs

```diff
@@ -0,0 +1,726 @@
+pub struct GlobalLocation(ModuleIndex, Loc);
+
+/// Type to represent diagnosis, like an error message.
+#[derive(Debug)]
+pub struct Diag {
+    /// A location for this diagnosis.
+    pub location: GlobalLocation,
+
+    /// The severity.
+    pub level: DiagLevel,
+
+    /// The message.
+    pub message: String,
+}
+
+/// The severity level of a diagnosis.
+#[derive(Debug)]
+pub enum DiagLevel {
+    ERROR,
+    WARNING,
+    HINT,
+}
+
+/// # Types
+
+/// The type declaration of a location. This is the same as `SignatureToken` except it is not scoped
+/// to a single module, and can contain Struct types coming from any of the modules in the
+/// environment.
+#[derive(Debug, Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub enum GlobalType {
+    Bool,
+    U8,
+    U64,
+    U128,
+    ByteArray,
+    Address,
+    Struct(ModuleIndex, StructDefinitionIndex, Vec<GlobalType>),
+    Reference(Box<GlobalType>),
+    MutableReference(Box<GlobalType>),
+    TypeParameter(TypeParameterIndex),
+}
+
+impl GlobalType {
+    pub fn is_reference(&self) -> bool {
+        match self {
+            GlobalType::Reference(_) | GlobalType::MutableReference(_) => true,
+            _ => false,
+        }
+    }
+}
+
+/// # Global Environment
+
+/// Global environment for a set of modules.
+#[derive(Debug)]
+pub struct GlobalEnv {
+    /// Options passed via the cli.
+    pub options: Options,
+
+    /// List of loaded modules, in order they have been provided using add().
+    pub module_data: Vec<ModuleData>,
+
+    // Accumulated diagnostics.
+    pub diags: RefCell<Vec<Diag>>,
+}
+
+impl GlobalEnv {
+    /// Creates a new environment.
+    pub fn new(options: Options) -> Self {
+        GlobalEnv {
+            options,
+            module_data: vec![],
+            diags: RefCell::new(vec![]),
+        }
+    }
+
+    /// Adds diagnosis to the environment.
+    pub fn add_diag(&self, loc: &GlobalLocation, level: DiagLevel, msg: &str) {
+        self.diags.borrow_mut().push(Diag {
+            location: *loc,
+            level,
+            message: msg.to_string(),
+        })
+    }
+
+    /// Creates data for a function, adding any information not contained in bytecode. This is
+    /// a helper for adding a new module to the environment.
+    pub fn create_function_data(
+        &self,
+        module: &VerifiedModule,
+        def_idx: FunctionDefinitionIndex,
+        arg_names: Vec<Identifier>,
+        type_arg_names: Vec<Identifier>,
+        spec: Vec<Condition>,
+    ) -> FunctionData {
+        let handle_idx = module.function_def_at(def_idx).function;
+        FunctionData {
+            def_idx,
+            handle_idx,
+            arg_names,
+            type_arg_names,
+            spec,
+        }
+    }
+
+    /// Creates data for a struct. Currently all information is contained in the byte code. This is
+    /// a helper for adding a new module to the environment.
+    pub fn create_struct_data(
+        &self,
```

Comment:
Unused?

Reply:
This and the next one are actually used in driver.rs.

Reply:
Ah, I meant the `&self` parameter (same below).